## Çoklu Doğrusal Regresyon Modeli Uygulaması:

Advertising veri setini kullanarak tv, radyo ve gazete reklamları ile bu reklamların satışa olan etkilerini inceleyeceğiz.  
Çoklu Doğrusal Regresyon Model'i kurarak bağımlı ve bağımsız değişkenlerimizin etkilerine inceleyeceğiz.

Bağımlı değişkenimiz 'sales'  
Bağımsız değişkenlerimiz ise 'tv, radyo ve gazete' değişkenleridir.  

Çoklu Doğrusal Regresyon Model'i, bir bağımlı ve birden fazla bağımsız değişkenden oluşmaktadır.  
Bu örnekte bağımsız değişkenlerimiz ise 'tv, radyo ve gazete' değişkenleridir.  


###  Örnek Bir Gerçek Hayat Projesinde Olması Gereken Adımlar:

1) Problem Tanımı ve Veri Toplama ===> Bir şirkette satışları (y) reklam harcamaları (TV, radyo, gazete) yardımıyla tahmin etmek istiyoruz.

2) Veriyi İnceleme ve Hazırlık ===> Veri Keşfi, Eksik ve Hatalı Verilerin Kontrolü

3) Bağımlı ve Bağımsız Değişkenleri Ayırma ===>  örn. =>  X = data[['TV', 'Radio', 'Newspaper']]  # Bağımsız değişkenler

4) Veriyi Eğitim ve Test Setlerine Bölme ===> X_train, X_test, y_train, y_test

5) Modeli Kurma ve Eğitme ===> model = LinearRegression()

6) Modelin Katsayılarını ve Denklemini İnceleme ===>    model.coef_ ve model.intercept_

7) Tahmin Yapmak => Test setindeki satışları tahmin edin: ===>  y_pred = model.predict(X_test)

8) Modelin Performansını Değerlendirme ===> Ortalama Kare Hatası (MSE) => mse = mean_squared_error(y_test, y_pred)

9) Sonuçları Yorumlama ===> 

10) Modeli İyileştirme ===>

In [4]:
import numpy as np
import pandas as pd

advertising = pd.read_csv("C:/Users/onurc/Documents/GitHub/dataScienceRepository/00_Data_Source/Advertising.csv")
df = advertising.copy()
df = df.iloc[:,1:len(df)]
df.head()

TV  radio  newspaper  sales
0  230.1   37.8       69.2   22.1
1   44.5   39.3       45.1   10.4
2   17.2   45.9       69.3    9.3
3  151.5   41.3       58.5   18.5
4  180.8   10.8       58.4   12.9

In [5]:
# Bağımlı ve bağımsız değişkenleri bir arada toplamak için atama işlemi

x = df.drop('sales', axis=1 )   # bağımsız değişkenler -- sales dışındakiler
y = df[['sales']]   # bağımlı değişken [[]]  dataFrame olması içindir.


In [6]:
x.head()

TV  radio  newspaper
0  230.1   37.8       69.2
1   44.5   39.3       45.1
2   17.2   45.9       69.3
3  151.5   41.3       58.5
4  180.8   10.8       58.4

In [7]:
y.head()

sales
0   22.1
1   10.4
2    9.3
3   18.5
4   12.9

### Modellemek için 2 farklı yöntem kullanacağız.

Bu şekilde yöntemlerin farklılıkları da ele alınacaktır.

1) StatsModel ile model kurmak :   
Doğrusal regresyon modeller hakkında daha fazla bilgi alabileceğimiz bir yapı elde edeceğiz.  

2) scikit-learn ile model kurmak :  
Sadece katsayılar ile ilgilendğimiz bir yapı elde edeceğiz.  
  
    (tercihen scikit-learn kullanılmaktadır.)

In [8]:
#Stats Model ile Kurmak

import statsmodels.api as sm

#OLS (Ordinary Least Squares) (Sıradan En Küçük Kareler)
#OLS, bağımlı bir değişken ile bir veya daha fazla bağımsız değişken arasındaki ilişkiyi modellemek için kullanılır.

lm = sm.OLS(y, x)
 
model = lm.fit()   


### Model Summary Açıklaması:   

Çıktı: OLS modelinin sonuç özetinde şunları bulabilirsiniz:

* R-squared (R Kare) = Bağımsız değişkenlerin, bağımlı değişkenleri açıklama yüzdesidir. (0.982 gayet iyi bir değerdir)  

* Adjusted. R-squared (Düzeltilmiş R Kare) = R Kare değerinin düzeltilmiş halidir. (Öznitelikten (özellik ekleme) dolayı şişmeyi önlemek için)  

* F-statistic: Modelin genel anlamlılık testini sağlar. (3566 oldukça yüksek bir F-istatistik değeridir = model genel olarak anlamlıdır.)

* Prob-F (P-value): Katsayıların anlamlılığını test eder.  (0.05'den oldukça küçük bir değer = model katsayıların anlamlı demektir. )

* Coefficients (Katsayılar): Bağımsız değişkenlerin model üzerindeki etkisini açıklar. 

* STD ERR (Standart Hata) : Bağımsız değişkenlerin standart hatasını açıklar.

* T-Statistic: Elimizdeki katsayıların anlamlılığını açıklar.  

* Prob-T-Statistics: Elimizdeki katsayıların olasılığını açıklar. (Prob-Olasılık)




In [9]:
model.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                                 OLS Regression Results                                
=======================================================================================
Dep. Variable:                  sales   R-squared (uncentered):                   0.982
Model:                            OLS   Adj. R-squared (uncentered):              0.982
Method:                 Least Squares   F-statistic:                              3566.
Date:                Thu, 19 Dec 2024   Prob (F-statistic):                   2.43e-171
Time:                        01:55:49   Log-Likelihood:                         -423.54
No. Observations:                 200   AIC:                                      853.1
Df Residuals:                     197   BIC:                                      863.0
Df Model:                           3                                                  
Covariance Type:            nonrobust                                                  
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
TV             0.0538      0.001     40.507      0.000       0.051       0.056
radio          0.2222      0.009     23.595      0.000       0.204       0.241
newspaper      0.0168      0.007      2.517      0.013       0.004       0.030
==============================================================================
Omnibus:                        5.982   Durbin-Watson:                   2.038
Prob(Omnibus):                  0.050   Jarque-Bera (JB):                7.039
Skew:                          -0.232   Prob(JB):                       0.0296
Kurtosis:                       3.794   Cond. No.                         12.6
==============================================================================

Notes:
[1] R² is computed without centering (uncentered) since the model does not contain a constant.
[2] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

### Katsayı Yorumlaması:

Diğer katsayılar sabit varsayılmak üzere

* Coef (TV) : TV reklamlarında 1 birimlik harcama, Satışlara ortalama 0.053 artış göstermektedir.

* Coef (Radio) : Radyo reklamlarında 1 birimlik harcama, Satışlara ortalama 0.22 artış göstermektedir.  

* Coef (Newspaper) : Gazete reklamlarında 1 birimlik harcama, Satışlara ortalama 0.016 artış göstermektedir.  

Radyo > TV > Gazete

### Gerçek Hayatta Kullanım:

Strateji Belirleme: Şirket, Radyo ve TV reklamlarına daha fazla bütçe ayırarak satışlarını artırmayı düşünebilir.  
Verimlilik Analizi: Gazete reklamlarının düşük etkisi göz önünde bulundurularak bu alandaki harcamalar azaltılabilir.  
Bu basit yorumlar, şirketin bütçe ve reklam stratejilerini optimize etmesine yardımcı olabilir.

### Scikit-Learn Kullanarak Model Kurmak

In [10]:
from sklearn.linear_model import LinearRegression

lm = LinearRegression()

model = lm.fit(x, y)


In [11]:
# Sabit değerimiz (b0)

model.intercept_

array([2.93888937])

In [12]:
# Kaysayılarımız (b1)

model.coef_

array([[ 0.04576465,  0.18853002, -0.00103749]])

In [13]:
# TV : 0.04576465
# Radio : 0.18853002
# Newspaper: 0.00103749

# Sabit değerimiz : 2.93888937

### Çoklu Doğrusal Regresyon Tahmini (Predict)

Sales = 2.94 + TV * 0.04 + Radio * 0.19 - Newspaper * 0.001 

In [14]:
# Çoklu Doğrusal Regresyon Fonksiyonu Gösterimi

# 30 Birim TV, 10 Birim Radyo, 40 Birim Gazete Harcaması yapılacaktır

2.94 + 30 * 0.04 + 10 * 0.19 - 40 * 0.001

5.999999999999999

In [15]:
harcamalar = pd.DataFrame([[30], [10], [40]]).T

In [16]:
harcamalar

0   1   2
0  30  10  40

In [17]:
model.predict(harcamalar)

c:\Users\onurc\anaconda3\Lib\site-packages\sklearn\base.py:464: UserWarning: X does not have valid feature names, but LinearRegression was fitted with feature names
  warnings.warn(


array([[6.15562918]])

## Kurulan Modelin Performansını Değerlendirme

MSE = Mean Squared Error = Ortalama Hata Karesi  
RMSE = Root Mean Squared Error = Ortalama Hata Karesi Karekökü


In [18]:
# MSE Import İşlemi

from sklearn.metrics import mean_squared_error


In [19]:
mean_squared_error

<function sklearn.metrics._regression.mean_squared_error(y_true, y_pred, *, sample_weight=None, multioutput='uniform_average', squared=True)>

In [20]:
y_true = y
y_pred = model.predict(x)

In [21]:
MSE = mean_squared_error(y_true, y_pred)

MSE

2.784126314510936

MSE = 2.784126314510936

Anlamı: Modelin, satış tahminlerinde ortalama olarak 2.784 birim karelik bir hata yaptığı anlamına gelir.  

Ancak, MSE’nin birimi, bağımlı değişkenin (satışların) biriminin karesi olduğu için doğrudan yorumlamak zordur.

In [22]:
RMSE = np.sqrt(MSE)

RMSE

1.6685701407225697

RMSE, MSE’nin karekökü alınarak bulunur ve bağımlı değişkenin birimiyle aynı birime sahiptir. Bu nedenle yorumlanması daha kolaydır.

Anlamı: Model, satışları ortalama olarak 1.67 birim hata ile tahmin ediyor.

### Hata Değeri Yorumu :

RMSE değeri ne kadar küçükse, model o kadar iyi tahmin yapıyor demektir.  
Burada 1.67 birimlik bir hata, bağımlı değişkenin (satışların) ölçeğine göre makul olup olmadığını değerlendirmeniz gerekir.   

Örneğin:  
Eğer satışlar genellikle 50-100 birim arasında değişiyorsa, 1.67 birimlik bir hata küçük bir hata olarak değerlendirilebilir.  
Ancak satışlar 5-10 birim arasında değişiyorsa, bu hata görece büyük bir hata olabilir.  

# Model Tuning (Model Doğrulama)

Çoklu doğrusal regresyon modelinde hiperparametre olmadığı için sadece model doğrulama gerçekleştireceğiz.  
Model doğrulama: Holdout, K-Cros validation... yöntemlerinin kullanımı

In [25]:
# Bağımsız değişkenler
x.head() 

TV  radio  newspaper
0  230.1   37.8       69.2
1   44.5   39.3       45.1
2   17.2   45.9       69.3
3  151.5   41.3       58.5
4  180.8   10.8       58.4

In [26]:
# Bağımlı değişkenler

y.head()

sales
0   22.1
1   10.4
2    9.3
3   18.5
4   12.9

# Holdout (Sınama Seti)

In [55]:
from sklearn.model_selection import train_test_split

In [74]:
# eğitim ve test setleri ikiye ayırma işlemi
# test_size = veri setinin %20'lik kısmını test için ayırmaktır (eğitim %80)

x_train, x_test, y_train, y_test = train_test_split(x, y, test_size = 0.20, random_state = 99)

In [75]:
x_train.head()

TV  radio  newspaper
16    67.8   36.6      114.0
51   100.4    9.6        3.6
97   184.9   21.0       22.0
164  117.2   14.7        5.4
71   109.8   14.3       31.7

In [76]:
y_train.head()

sales
16    12.5
51    10.7
97    15.5
164   11.9
71    12.4

In [77]:
x_test.head()

TV  radio  newspaper
135  48.3   47.0        8.5
127  80.2    0.0        9.2
191  75.5   10.8        6.0
66   31.5   24.6        2.2
119  19.4   16.0       22.3

In [78]:
y_test.head()

sales
135   11.6
127    8.8
191    9.9
66     9.5
119    6.6

In [79]:
lm = LinearRegression()
model = lm.fit(x_train, y_train)

In [80]:
# Eğitim (Train) Hatası
np.sqrt(mean_squared_error(y_train, model.predict(x_train)))

1.7236824822650754

In [82]:
# Test Hatası
np.sqrt(mean_squared_error(y_test, model.predict(x_test)))

1.431278313830164

## K-Katlı Çapraz Doğrulama Yöntemi (K-Fold Cross Validation Method)

In [83]:
from sklearn.model_selection import cross_val_score

In [91]:
cross_val_score(model, x_train, y_train, cv = 10, scoring = "neg_mean_squared_error")

array([-2.1019073 , -2.48953197, -3.09704214, -2.34694216, -3.68175761,
       -1.8691401 , -3.18173007, -4.1927349 , -2.17128376, -8.03821974])

In [94]:
# Eğitim setinin ortalama hatası 3.31

# MSE (Hata Kareler Ortalaması)

np.mean(-cross_val_score(model, x_train, y_train, cv = 10, scoring = "neg_mean_squared_error"))

3.317028974234124

In [95]:
# RMSE (Hata Kareler Ortalaması Karekökü)

np.sqrt(np.mean(-cross_val_score(model, x_train, y_train, cv = 10, scoring = "neg_mean_squared_error")))

1.821271252239524